Add "Show outliers" option

openjck opened the pull request.

openjck commented:
This works, but it's pretty slow. Still looking into it...

openjck commented:
Updated and ready for review

openjck commented:
This won't do much on charts with few data points; I recommend using prod sample data and testing the _bookmarksAtMostRecentDay_ chart.

spasovski commented on distributionviewer/core/static/js/app/components/views/chart-detail.js:
```diff
@@ -4,10 +4,21 @@ import ChartContainer from '../containers/chart-container';
 export class ChartDetail extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {showOutliers: false};
```
I agree with this as long as it's obvious enough in the UI.

openjck replied:
Sure thing. This is also the default in Brendan's dashboard.

spasovski commented on distributionviewer/core/static/js/app/components/views/chart.js:
```diff
@@ -102,10 +50,102 @@ export class Chart extends React.Component {
+    } else {
+      this.setState({
+        min_x: this.pointsMeta[Math.max(Math.round(pointsMetaLength * .005) - 1, 0)].x,
+        max_x: this.pointsMeta[Math.min(Math.round(pointsMetaLength * .995) - 1, pointsMetaLength)].x,
```
The math here is fishy for sure. My understanding is we'll need to figure out the standard deviation and decide to drop any values that exceed it (though technically our request is to use the central 99%). Once we pop those from the array we wouldn't need to mess with min_x, max_x etc.

If the idea really is to just use the "central 99%" we'll need to sort the values in the data (only if it exceeds 99 entries) and then drop the head and tail ends as needed. I can recommend a brand of bourbon that might help with these functions.

spasovski commented:
I like the component changes but we should rethink the math as well as whether we want to keep regenerating the whole chart. The option should also not be available for any data that doesn't exceed some length threshold like 100.

Honestly this is not a simple task for the project and it might be a good idea to go over this with product to see whether it's even worth it. I'd rather they pick a standard deviation threshold and have us drop any data points that surpass it.

openjck commented:
Updated the math. Still thinking about the best way to only show the _Show outliers_ option for non-category charts. The chart detail page doesn't know what kind of chart it's going to show, and it would be wasteful to do another request just to find out.

spasovski commented:
r+pc (I'd like to see product's opinion on this feature first). Thanks for the refactoring work too looks good.

spasovski commented on distributionviewer/core/static/js/app/components/views/chart.js:
```diff
@@ -102,10 +50,101 @@ export class Chart extends React.Component {
+    const pointsMetaLength = this.pointsMeta.length;
+    let minIndex, maxIndex;
+
+    if (showOutliers) {
```
The math in the `else` would account for it but we could avoid it for smaller data sets by doing `|| pointsMetaLength < 100`. Treat this as a nitpick - it's not necessary.

spasovski commented:
The product folks seem happy with this option as-is so merge when ready.

openjck commented:
> The product folks seem happy with this option as-is so merge when ready.

Sure thing. I'll just disable the toggle on non-applicable charts first.

openjck committed:
Add "Show outliers" option

openjck commented:
Updated and rebased! The checkbox isn't the prettiest thing ever (I prefer the toggle you made for SHIELDDash) but we can follow up with that. The supporting code changes are big enough already.

To test, I recommend using prod endpoints and checking out the _bookmarksAtMostRecentDay_ chart.

spasovski commented on distributionviewer/core/static/js/app/components/containers/chart-detail-container.js:
```diff
@@ -1,10 +1,20 @@
-  return (
-    <ChartDetail {...props}  />
-  );
+const mapStateToProps = function(store) {
```
You're right we should've done this earlier. Good call.

spasovski commented:
Looks good but we lost the old prod data. r+ in any case thanks for handling this.
